Rely on ReportConfig object to populate saved UCR filters

gherceg opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[JIRA Ticket](https://dimagi-dev.atlassian.net/browse/SAAS-12059)

This originally came up because some projects have UCRs with lots of filters and/or filter options. Since all of these filter user selections are specified via URL parameters, the URL can end up being too large to access, resulting in a 414 status code response. This is specifically an issue for saved reports because the typical report workflow passes filter options via json in a POST method (think clicking Apply on some report).

Since we already store filter information for saved UCRs in the `ReportConfig` object, we should use this to populate the filters that are rendered in the HTML on HQ's side. This PR targets saved UCRs specifically, replacing the UCR view's property`filter_context` in the HTML context with the filters from the saved report. This is what is used to populate the filters [here](https://github.com/dimagi/commcare-hq/blob/7df6a8823d4160308ab537f8490969e4210e8686/corehq/apps/reports_core/templates/reports_core/partials/filter_panel.html#L15).

I admittedly don't have the most in-depth knowledge of reports, so any feedback is much appreciated. There are definitely other ways to achieve this result and I'd like to know if there are better ones. Planning on requesting QA once this is more finalized.
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
All saved UCR reports will only need the config_id specified in the URL to render accurately. Existing URLs will continue to work, but any newly rendered hyperlinks to a saved UCR in HQ will have the new shortened url. 
## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added very minor unit tests to verify that `query_string` returns the expected result for a saved configurable report and a saved non-configurable report.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
[QA Ticket](https://dimagi-dev.atlassian.net/browse/QA-2975)
### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Locally tested and tested on staging. This will only impact UCRs (not that that is a small blast radius).
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

gherceg committed:
rely on config id to populate filters

gherceg committed:
filter should handle date objects from saved report filters

gherceg committed:
minor test for query_string

esoergel commented on corehq/apps/userreports/reports/view.py:
```diff
@@ -392,6 +403,12 @@ def _get_context_for_saved_report(report_config):
 
+        if saved_report_config:
+            # override filter_context with the saved report config filters
+            context.update({'filter_context': self.filter_context_for_saved_report(saved_report_config)})
```
Do you know if this is actually necessary?  Know what `default_config` is for if not rendering the filters?  Maybe pre-populating the "save report" modal?

gherceg replied:
Yeah I should have specified this up above because `default_config` is used client-side to update saved report filters right now, but it doesn't work. There are a couple of reasons I went with this approach instead of using `default_config` to update the DOM client side. 
1) There's a specific format for filter values (`{"value": <value>,  "display": <display>}`) which the client side didn't know about (unless this standard across HQ in which case this point is moot. 
2) The changes needed client side were starting to feel unnecessarily complex. Right now this code exists but does not work:
https://github.com/dimagi/commcare-hq/blob/1b34ad8f49a281e5bdf146127bf7a544dc4245bd/corehq/apps/reports/static/reports/js/report_config_models.js#L140-L146
We would need to change `$element.val()` to `$element.attr('data-initial-values', <correctly formatted dict>)`, keeping in mind that format is not something client side is aware of, and would have to add another case for Datespan filters since they have a different attribute to update/different format to be aware of.

esoergel commented on corehq/apps/userreports/reports/view.py:
```diff
@@ -369,6 +375,11 @@ def pop_report_builder_context_data(self):
+        """
+        Update context with saved report data
+        Override filter_context with saved report config filters
+        :return: updated context
```
nit: this doesn't actually update the context - it returns some context that will be inserted into the main context.

gherceg replied:
Fair point

esoergel commented on corehq/apps/userreports/reports/view.py:
```diff
@@ -255,8 +255,14 @@ def filter_values(self):
+            for report_filter in self.filters
+        }
+
+    def filter_context_for_saved_report(self, saved_config):
```
what would you think about moving this inside `filter_context`?  That way `self.filter_context` always returns the correct value, and is always applicable.  To do so, you'd probably need to pull the call that gets `saved_report_config` (lines 391 and 392) into a `memoized` or `@cached_property` so it can be referenced in both places.

gherceg replied:
Yeah I like that. I made that change on the latest commit.

czue reviewed (commented):
One potential gotcha to look at is what the behavior is when you load a saved report, then mess with the filters after it loads. When I looked into [this other ticket](https://dimagi-dev.atlassian.net/browse/SAAS-11248) a while back that behavior was basically "a mess" but it'd be good to at least understand if/how it changes with this update.

I admit I don't really have my head wrapped around this code at all, which it makes it hard for me to stamp that this is the right thing to do (though it makes logical sense to me).

gherceg commented:
Just realized I have a lot more testing to do on staging which will probably result in some code updates so reviewers can hold off for now.

gherceg committed:
Merge branch 'master' into gh/remove-filters-from-saved-ucrs

gherceg commented on corehq/apps/reports_core/filters.py:
```diff
@@ -139,7 +139,9 @@ def value(self, **kwargs):
             if param:
-                if self.compare_as_string:
+                if isinstance(param, date):
+                    return datetime.combine(param, time())
```
This is an artifact of passing in filters directly from the saved report configuration rather than url parameters.

gherceg replied:
@czue I got very distracted but would like to wrap this up. After picking this back up, I realized this is not a very elegant solution. Would you agree that instead of adding an additional case to the DatespanFilter to accept actual datetime objects, the optimal solution would be to return a properly formatted dictionary from saved_report_config here:

https://github.com/dimagi/commcare-hq/blob/2ddc6e48a27353c1bf7e0c002cec5801e5f1e840/corehq/apps/userreports/reports/view.py#L256-L261

czue replied:
Ah I see. Just to confirm my understanding:

1. Prior to this change it always returned a `request_dict` that had serialized data
2. The code that was modified in this PR changes it to return a native (couch?) dict that has python objects in it.
3. You're proposing changing the new code to serialize the filters before being returned, so it looks like a `request_dict`?

That makes sense to me. Wondering if there's a better name for "get me the serialized version of the filters" than `filter_context` but changing the name is probably out of scope unless easy.

millerdev reviewed (commented):
Didn't do a thorough review yet, but had one nitpick after a quick review.

gherceg committed:
code review edits

- remove boilerplate code from tests
- make saved_report_config a cached property so filter_context can accomodate
  both request params and config params

gherceg commented on corehq/apps/saved_reports/tests/test_saved_report_urls.py:
```diff
@@ -0,0 +1,55 @@
+
+    @classmethod
+    def tearDownClass(cls):
+        super(SavedReportURLTest, cls).tearDownClass()
```
Removed ✅

gherceg commented:
So I tested the behavior @czue suggested with changing filters after loading, as well as loading configs from Favorites. Both cases seem to work as expected for me. 

Reading through the ticket Cory linked, it makes sense that this fixes the Favorites issue because @mkangia had mentioned when loading a config from the Favorites section, url parameters were not specified, only the config_id. I can confirm that prior to this PR, only specifying the `config_id` does not update filters (though it looks like it worked at one point).  

If this seems like the best approach to move forward with, I will also look into cleaning up the broken JS code that attempts to set filters client side, as well as potentially returning only data that is needed in the response of this request.

esoergel commented on corehq/apps/userreports/reports/view.py:
```diff
@@ -254,14 +254,9 @@ def filter_values(self):
     @property
     @memoized
     def filter_context(self):
```
:+1:

gherceg committed:
improve tests; remove no longer necessary js code

gherceg committed:
add back that unnecessary js code...

gherceg committed:
pass correctly formatted data into report filters

gherceg committed:
update filters before setting href

gherceg committed:
stickler

gherceg commented:
This turned out to be a bit more messy than I had hoped so sorry about that. I view this as solving the immediate problem, and with a potential followup PR to address a better method for populating filters. This still depends on JS code to update the DOM specifically for datespan filters which is fine. Requesting review from @czue, @millerdev, and @esoergel.

czue commented on corehq/apps/saved_reports/models.py:
```diff
@@ -241,6 +241,22 @@ def query_string(self):
+    @memoized
+    def serialized_filters(self):
+        """
+        converts any filter values that are objects to strings
```
nit: this doc string is slightly misleading. i think it only converts dates, right?

czue commented on corehq/apps/saved_reports/tests/test_saved_report_model.py:
```diff
@@ -0,0 +1,31 @@
+class SavedReportModelTest(TestCase):
+
+    def test_saved_report_serialized_filters(self):
+        # NOTE:
```
:thinking:

gherceg replied:
Thanks for flagging. Wanted to leave a note that the date values in `filters` would be converted to date objects, otherwise this test is pretty lame. Instead of a comment I added a couple of assertions to convey what I wanted.

gherceg committed:
cleanup comments and test
